p2panda-net/src/sync/tests.rs

```diff
@@ -0,0 +1,456 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+mod sync_protocols {
+    use std::sync::Arc;
+
+    use async_trait::async_trait;
+    use futures_lite::{AsyncRead, AsyncWrite, StreamExt};
+    use futures_util::{Sink, SinkExt};
+    use p2panda_sync::cbor::{into_cbor_sink, into_cbor_stream};
+    use p2panda_sync::{FromSync, SyncError, SyncProtocol};
+    use serde::{Deserialize, Serialize};
+
+    use super::TestTopic;
+
+    #[derive(Debug, Serialize, Deserialize)]
+    enum ProtocolMessage {
+        Topic(TestTopic),
+        Done,
+    }
+
+    /// A sync implementation which returns a mocked error.
+    #[derive(Debug)]
+    pub enum FailingProtocol {
+        /// A critical error is triggered inside `accept()` after sync messages have been
+        /// exchanged.
+        AcceptorFailsCritical,
+
+        /// A critical error is triggered inside `initiate()` after the handshake is complete.
+        InitiatorFailsCritical,
+
+        /// An unexpected behaviour error is triggered inside `accept()` by sending the topic twice
+        /// from `initiate()`.
+        InitiatorSendsTopicTwice,
+
+        /// An unexpected behaviour error is triggered inside `initiate()` by sending a topic from
+        /// `accept()`.
+        AcceptorSendsTopic,
+
+        /// No errors are explicitly triggered; used for "happy path" test.
+        NoError,
+    }
+
+    #[async_trait]
+    impl<'a> SyncProtocol<'a, TestTopic> for FailingProtocol {
+        fn name(&self) -> &'static str {
+            "failing-protocol"
+        }
+
+        async fn initiate(
+            self: Arc<Self>,
+            topic: TestTopic,
+            tx: Box<&'a mut (dyn AsyncWrite + Send + Unpin)>,
+            rx: Box<&'a mut (dyn AsyncRead + Send + Unpin)>,
+            mut app_tx: Box<
+                &'a mut (dyn Sink<FromSync<TestTopic>, Error = SyncError> + Send + Unpin),
+            >,
+        ) -> Result<(), SyncError> {
+            let mut sink = into_cbor_sink(tx);
+            let mut stream = into_cbor_stream(rx);
+
+            sink.send(ProtocolMessage::Topic(topic.clone())).await?;
+
+            // Simulate critical sync implementation bug by sending the topic a second time.
+            if let FailingProtocol::InitiatorSendsTopicTwice = *self {
+                sink.send(ProtocolMessage::Topic(topic.clone())).await?;
+            }
+
+            // Simulate some critical error which occurred inside the sync session.
+            if let FailingProtocol::InitiatorFailsCritical = *self {
+                return Err(SyncError::Critical(
+                    "something really bad happened in the initiator".to_string(),
+                ));
+            }
```

Comment:
I've moved the critical failure higher in the protocol so we can test what happens if the connection suddenly disappears.

Reply:
I really like the way you separated out the helper function for running each protocol impl; it makes the actual tests very easy to reason about.